File: back/src/Liane/Liane.Api/Notification/Notification.cs
using System;
using Liane.Api.Util.Ref;

namespace Liane.Api.Notification;

public sealed record Notification(
  string title,
  string message,
  NotificationPayload Payload
);
public abstract record NotificationPayload(
  string? Id,
  bool Seen,
  DateTime CreatedAt) : IIdentity
{
  public sealed record WithEvent<T>(
    string? Id,
    DateTime CreatedAt,
    T Event,
    bool Seen = false
  ) : NotificationPayload(Id, Seen, CreatedAt) where T : class
  {
    public string Type => typeof(T).Name;
  }
}
  

File: app/src/api/notification.ts
import { ConversationGroup, Ref, User, UTCDateTime } from "@/api/index";
import { LianeEvent } from "@/api/event";

export type Notification = (Info | Event | NewMessage) & AbstractNotification;

export enum Answer {
  Accept = "Accept",
  Reject = "Reject"
}

export type Recipient = Readonly<{
  user: Ref<User>;
  seenAt?: UTCDateTime;
}>;

type AbstractNotification = Readonly<{
  type: string;
  id?: string;
  createdBy: Ref<User>;
  createdAt: UTCDateTime;
  recipients: Recipient[];
  answers: Answer[];
  title: string;
  message: string;
}>;

export type Info = Readonly<{
  type: "Info";
}> &
  AbstractNotification;
/*
export type Reminder = Readonly<{
  type: "Reminder";
  payload: {
    liane: Ref<Liane>;
    trip: WayPoint[];
    //   rallyingPoint: RallyingPoint;
    //  at: UTCDateTime;
  };
}> &
  AbstractNotification;*/

export type NewMessage = Readonly<{
  type: "NewMessage";
  conversation: Ref<ConversationGroup>;
}> &
  AbstractNotification;

export type Event<T extends LianeEvent = LianeEvent> = Readonly<{
  type: "Event";
  payload: T;
}> &
  AbstractNotification;
